flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -363,1691 +372,1780 @@ public int getKeyGroupPrefixBytes() {
 
-				// TODO replace with state migration; note that key hash codes need to remain the same after migration
-				throw new StateMigrationException("The new key serializer is not compatible to read previous keys. " +
-					"Aborting now since state migration is currently not available");
+				if (heap.isEmpty()) {
+					currentSubIterator = null;
+					valid = false;
+				} else {
+					currentSubIterator = heap.poll();
+					newKVState = true;
+					detectNewKeyGroup(oldKey);
+				}
 			}
+		}
 
-			this.keygroupStreamCompressionDecorator = serializationProxy.isUsingKeyGroupCompression() ?
-				SnappyStreamCompressionDecorator.INSTANCE : UncompressedStreamCompressionDecorator.INSTANCE;
-
-			List<RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?>> restoredMetaInfos =
-				serializationProxy.getStateMetaInfoSnapshots();
-			currentStateHandleKVStateColumnFamilies = new ArrayList<>(restoredMetaInfos.size());
-			//rocksDBKeyedStateBackend.restoredKvStateMetaInfos = new HashMap<>(restoredMetaInfos.size());
+		private boolean isDifferentKeyGroup(byte[] a, byte[] b) {
+			return 0 != compareKeyGroupsForByteArrays(a, b, keyGroupPrefixByteCount);
+		}
 
-			for (RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?> restoredMetaInfo : restoredMetaInfos) {
+		private void detectNewKeyGroup(byte[] oldKey) {
+			if (isDifferentKeyGroup(oldKey, currentSubIterator.currentKey)) {
+				newKeyGroup = true;
+			}
+		}
 
-				Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>> registeredColumn =
-					rocksDBKeyedStateBackend.kvStateInformation.get(restoredMetaInfo.getName());
+		/**
+		 * Returns the key-group for the current key.
+		 * @return key-group for the current key
+		 */
+		public int keyGroup() {
+			int result = 0;
+			//big endian decode
+			for (int i = 0; i < keyGroupPrefixByteCount; ++i) {
+				result <<= 8;
+				result |= (currentSubIterator.currentKey[i] & 0xFF);
+			}
+			return result;
+		}
 
-				if (registeredColumn == null) {
-					byte[] nameBytes = restoredMetaInfo.getName().getBytes(ConfigConstants.DEFAULT_CHARSET);
+		public byte[] key() {
+			return currentSubIterator.getCurrentKey();
+		}
 
-					ColumnFamilyDescriptor columnFamilyDescriptor = new ColumnFamilyDescriptor(
-						nameBytes,
-						rocksDBKeyedStateBackend.columnOptions);
+		public byte[] value() {
+			return currentSubIterator.getIterator().value();
+		}
 
-					RegisteredKeyedBackendStateMetaInfo<?, ?> stateMetaInfo =
-						new RegisteredKeyedBackendStateMetaInfo<>(
-							restoredMetaInfo.getStateType(),
-							restoredMetaInfo.getName(),
-							restoredMetaInfo.getNamespaceSerializer(),
-							restoredMetaInfo.getStateSerializer());
+		/**
+		 * Returns the Id of the k/v state to which the current key belongs.
+		 * @return Id of K/V state to which the current key belongs.
+		 */
+		public int kvStateId() {
+			return currentSubIterator.getKvStateId();
+		}
 
-					rocksDBKeyedStateBackend.restoredKvStateMetaInfos.put(restored
```

Comment:
I skimmed over the changes and I assume that you've mostly copied the code from one place to another without bigger changes. But it's really hard to see any real changes due to the combined movement of the code.